microhapulator/qcsummary.py

```diff
@@ -0,0 +1,84 @@
+# -------------------------------------------------------------------------------------------------
+# Copyright (c) 2024, DHS.
+#
+# This file is part of MicroHapulator (https://github.com/bioforensics/microhapulator) and is
+# licensed under the BSD license: see LICENSE.txt.
+#
+# This software was prepared for the Department of Homeland Security (DHS) by the Battelle National
+# Biodefense Institute, LLC (BNBI) as part of contract HSHQDC-15-C-00064 to manage and operate the
+# National Biodefense Analysis and Countermeasures Center (NBACC), a Federally Funded Research and
+# Development Center.
+# -------------------------------------------------------------------------------------------------
+
+from .qcstats import FilterStats, PairedAmbiguityFilterStats, ReadMergingStats
+from dataclasses import dataclass
+
+
+@dataclass
+class SingleEndReadQCSummary:
+    ambig: FilterStats
+    length: FilterStats
+
+    @property
+    def total(self):
+        return self.ambig.filtered + self.ambig.retained
+
+    @property
+    def total_reads(self):
+        return f"{self.total:,}"
+
+    @property
+    def filtered_ambig(self):
+        return SingleEndReadQCSummary.render(self.ambig.filtered, self.total)
+
+    @property
+    def filtered_length(self):
+        return SingleEndReadQCSummary.render(self.length.filtered, self.total)
+
+    @property
+    def retention(self):
+        return SingleEndReadQCSummary.render(self.length.retained, self.total)
+
+    @staticmethod
+    def render(numerator, denominator):
+        rate = numerator / denominator * 100
+        return f"{numerator:,} ({rate:.1f}%)"
+
+    @classmethod
+    def from_workdir(cls, wdpath, sample):
+        ambig_path = f"{wdpath}/analysis/{sample}/{sample}-ambig-read-counts.txt"
+        length_path = f"{wdpath}/analysis/{sample}/{sample}-length-filtered-read-counts.txt"
+        ambig_stats = FilterStats.from_txt(ambig_path)
+        length_stats = FilterStats.from_txt(length_path)
+        return cls(ambig_stats, length_stats)
+
+    @staticmethod
+    def collect(samples, workdir="."):
+        sampleqc = dict()
+        for sample in sorted(samples):
+            sampleqc[sample] = SingleEndReadQCSummary.from_workdir(workdir, sample)
+        return sampleqc
+
+
+@dataclass
+class PairedReadQCSummary:
+    ambig: PairedAmbiguityFilterStats
+    merge: ReadMergingStats
+    length: FilterStats
```

Comment:
For paired-end data, the simple `FilterStats` data structure works for length filtering post-merge, but the ambiguity filtering and read merge stats needed their own dedicated data structures. This class is essentially a container for the three data structures, along with convenience methods for loading from the working directory.